In [2]:
#Manually add the Summer Olympics, London 2012 dataset as arrays'

country=np.array(['GBR','CHN','RUS','US','KOR','JPN','GER'])
gold=np.array([29, 38, 24, 46, 13, 7, 11])
silver=np.array([17, 28, 25, 28, 8, 14, 11])
bronze=np.array([19, 22, 32, 29, 7, 17, 14])



In [6]:
#Use the argmax() method to find the highest number of gold medals
gold_max_ind=gold.argmax()
gold_max_ind

3

In [4]:
#Print the name of the country
country[gold_max_ind]

'US'

In [5]:
#Use a for loop to create the required output
for n,i in enumerate(country):
    Country=i
    Gold=gold[n]
    total=Gold + silver[n] +bronze[n]
    print('{} has {} gold medals and {} medals totally'.format(Country, Gold, total))
#Alternatively
'''for i in range(len(country)):
    Gold=gold[i]
    Country=country[i]
    total=bronze[i]+gold[i]+silver[i]
    print('{} has {} gold medals and {} medals totally'.format(Country, Gold, total))'''

GBR has 29 gold medals and 65 medals totally
CHN has 38 gold medals and 88 medals totally
RUS has 24 gold medals and 81 medals totally
US has 46 gold medals and 103 medals totally
KOR has 13 gold medals and 28 medals totally
JPN has 7 gold medals and 38 medals totally
GER has 11 gold medals and 36 medals totally


"for i in range(len(country)):\n    Gold=gold[i]\n    Country=country[i]\n    total=bronze[i]+gold[i]+silver[i]\n    print('{} has {} gold medals and {} medals totally'.format(Country, Gold, total))"

In [7]:
#Manually add the dataset

countries=np.array(['Algeria','Angola','Argentina','Australia','Austria','Bahamas','Bangladesh','Belarus',
                    'Belgium','Bhutan','Brazil','Bulgaria','Cambodia','Cameroon','Chile','China','Colombia',
                    'Cyprus','Denmark','El Salvador','Estonia','Ethiopia','Fiji','Finland','France','Georgia',
                    'Ghana','Grenada','Guinea','Haiti','Honduras','Hungary','India','Indonesia','Ireland','Italy',
                    'Japan','Kenya', 'South Korea','Liberia','Malaysia','Mexico', 'Morocco','Nepal','New Zealand',
                    'Norway','Pakistan', 'Peru','Qatar','Russia','Singapore','South Africa','Spain','Sweden',
                    'Switzerland','Thailand', 'United Arab Emirates','United Kingdom','United States','Uruguay',
                    'Venezuela','Vietnam','Zimbabwe'])




each_country_gdp=np.array([2255.225482,629.9553062,11601.63022,25306.82494,27266.40335,19466.99052,588.3691778,
                           2890.345675,24733.62696,1445.760002,4803.398244,2618.876037,590.4521124,665.7982328,
                           7122.938458,2639.54156,3362.4656,15378.16704,30860.12808,2579.115607,6525.541272,
                           229.6769525,2242.689259,27570.4852,23016.84778,1334.646773,402.6953275,6047.200797,
                           394.1156638,385.5793827,1414.072488,5745.981529,837.7464011,1206.991065,27715.52837,
                           18937.24998,39578.07441,478.2194906,16684.21278,279.2204061,5345.213415,6288.25324,
                           1908.304416,274.8728621,14646.42094,40034.85063,672.1547506,3359.517402,36152.66676,
                           3054.727742,33529.83052,3825.093781,15428.32098,33630.24604,39170.41371,2699.123242,
                           21058.43643,28272.40661,37691.02733,9581.05659,5671.912202,757.4009286,347.7456605])

In [8]:
#Use the argmax() method to find the highest GDP
#argmax() returns index of value with max
max_gdp=each_country_gdp.argmax()
max_gdp

45

In [9]:
#Print the name of the country
country_max_gdp=countries[max_gdp]
country_max_gdp

'Norway'

In [10]:
max_gdp_value=each_country_gdp.max()
max_gdp_value

40034.85063

In [11]:
#Use the argmin() method to find the lowest GDP
min_gdp=each_country_gdp.argmin()
min_gdp

21

In [13]:
#Print the name of the country
min_gdp_country=countries[21]
min_gdp_country

'Ethiopia'

In [14]:
min_gdp_value=each_country_gdp.min()
min_gdp_value

229.6769525

In [15]:
#Use a for loop to print the required output
for country in countries:
    print('evaluating country '+ country)
    # or print('evaluating country {}'.format(country))

evaluating country Algeria
evaluating country Angola
evaluating country Argentina
evaluating country Australia
evaluating country Austria
evaluating country Bahamas
evaluating country Bangladesh
evaluating country Belarus
evaluating country Belgium
evaluating country Bhutan
evaluating country Brazil
evaluating country Bulgaria
evaluating country Cambodia
evaluating country Cameroon
evaluating country Chile
evaluating country China
evaluating country Colombia
evaluating country Cyprus
evaluating country Denmark
evaluating country El Salvador
evaluating country Estonia
evaluating country Ethiopia
evaluating country Fiji
evaluating country Finland
evaluating country France
evaluating country Georgia
evaluating country Ghana
evaluating country Grenada
evaluating country Guinea
evaluating country Haiti
evaluating country Honduras
evaluating country Hungary
evaluating country India
evaluating country Indonesia
evaluating country Ireland
evaluating country Italy
evaluating country Japan
evalu

In [16]:
#Use a for loop to print the required list

'''for i in range(len(countries)):
    country=countries[i]
    gdp=each_country_gdp[i]
    print('{} has {} GDP'.format(country, gdp))'''


for country, gdp in zip(countries, each_country_gdp):
    print('{} has a GDP of {}'.format(country, gdp))
    

Algeria has a GDP of 2255.225482
Angola has a GDP of 629.9553062
Argentina has a GDP of 11601.63022
Australia has a GDP of 25306.82494
Austria has a GDP of 27266.40335
Bahamas has a GDP of 19466.99052
Bangladesh has a GDP of 588.3691778
Belarus has a GDP of 2890.345675
Belgium has a GDP of 24733.62696
Bhutan has a GDP of 1445.760002
Brazil has a GDP of 4803.398244
Bulgaria has a GDP of 2618.876037
Cambodia has a GDP of 590.4521124
Cameroon has a GDP of 665.7982328
Chile has a GDP of 7122.938458
China has a GDP of 2639.54156
Colombia has a GDP of 3362.4656
Cyprus has a GDP of 15378.16704
Denmark has a GDP of 30860.12808
El Salvador has a GDP of 2579.115607
Estonia has a GDP of 6525.541272
Ethiopia has a GDP of 229.6769525
Fiji has a GDP of 2242.689259
Finland has a GDP of 27570.4852
France has a GDP of 23016.84778
Georgia has a GDP of 1334.646773
Ghana has a GDP of 402.6953275
Grenada has a GDP of 6047.200797
Guinea has a GDP of 394.1156638
Haiti has a GDP of 385.5793827
Honduras has a 

Highest GPD value---.max()

Lowest GDP value----.min()

Mean GDP value---.mean()

Standardized GDP value----.std()

Sum of all the GDPs----.sum()

# Series

In [1]:
#creating series.....
import pandas as pd
import numpy as np


In [13]:
list= ['a','b','c','d','e']

In [14]:
Sr = pd.Series(list)
Sr

0    a
1    b
2    c
3    d
4    e
dtype: object

In [15]:
np_country=np.array(['lux','Norway','Japan','Switz','US','Qarta','Iceland','Swede','Singapore','Denmark'])

In [17]:
Sra=pd.Series(np_country)
Sra

0          lux
1       Norway
2        Japan
3        Switz
4           US
5        Qarta
6      Iceland
7        Swede
8    Singapore
9      Denmark
dtype: object

In [19]:
#we can set a key like name,Id etc as index

dict_gdp=pd.Series([2255.225482,629.9553062,11601.63022,25306.82494,27266.40335,19466.99052,588.3691778,
                           2890.345675,24733.62696,1445.760002,4803.398244,2618.876037,590.4521124,665.7982328,
                           7122.938458,2639.54156,3362.4656,15378.16704,30860.12808,2579.115607,6525.541272,
                           229.6769525,2242.689259,27570.4852,23016.84778,1334.646773,402.6953275,6047.200797,
                           394.1156638,385.5793827,1414.072488,5745.981529,837.7464011,1206.991065,27715.52837,
                           18937.24998,39578.07441,478.2194906,16684.21278,279.2204061,5345.213415,6288.25324,
                           1908.304416,274.8728621,14646.42094,40034.85063,672.1547506,3359.517402,36152.66676,
                           3054.727742,33529.83052,3825.093781,15428.32098,33630.24604,39170.41371,2699.123242,
                           21058.43643,28272.40661,37691.02733,9581.05659,5671.912202,757.4009286,347.7456605],
                    index=['Algeria','Angola','Argentina','Australia','Austria','Bahamas','Bangladesh','Belarus',
                    'Belgium','Bhutan','Brazil','Bulgaria','Cambodia','Cameroon','Chile','China','Colombia',
                    'Cyprus','Denmark','El Salvador','Estonia','Ethiopia','Fiji','Finland','France','Georgia',
                    'Ghana','Grenada','Guinea','Haiti','Honduras','Hungary','India','Indonesia','Ireland','Italy',
                    'Japan','Kenya', 'South Korea','Liberia','Malaysia','Mexico', 'Morocco','Nepal','New Zealand',
                    'Norway','Pakistan', 'Peru','Qatar','Russia','Singapore','South Africa','Spain','Sweden',
                    'Switzerland','Thailand', 'United Arab Emirates','United Kingdom','United States','Uruguay',
                    'Venezuela','Vietnam','Zimbabwe'])
dict_gdp.head(5)

Algeria       2255.225482
Angola         629.955306
Argentina    11601.630220
Australia    25306.824940
Austria      27266.403350
dtype: float64

In [22]:
scalar_s=pd.Series(5,index=list)
scalar_s

a    5
b    5
c    5
d    5
e    5
dtype: int64

## Accessing elements in Series

In [23]:
dict_gdp[0]

2255.225482

In [24]:
dict_gdp[0:5]

Algeria       2255.225482
Angola         629.955306
Argentina    11601.630220
Australia    25306.824940
Austria      27266.403350
dtype: float64

In [27]:
dict_gdp.loc['Qatar']

36152.66676

In [28]:
dict_gdp.iloc[0]

2255.225482

## Vectorized operations in series

In [29]:
fs=pd.Series([1,2,3,4],index=['a','b','c','d'])
ss=pd.Series([10,20,30,40],index=['a','b','c','d'])
fs+ss

a    11
b    22
c    33
d    44
dtype: int64

In [30]:
#we get Nan if index doesnt coincide
os=pd.Series([10,20,30,40],index=['a','d','c','b'])
fs+os

a    11
b    42
c    33
d    24
dtype: int64

# DataFrame

## creating Dataframe

In [32]:
dict_olymp={'HostCity':['GBR','CHN','RUS','US','KOR'],'Year':[2012,2008,2004,2000,2006],
          'No of countries':[205,204,201,200,197]}
dict_olymp


{'HostCity': ['GBR', 'CHN', 'RUS', 'US', 'KOR'],
 'Year': [2012, 2008, 2004, 2000, 2006],
 'No of countries': [205, 204, 201, 200, 197]}

In [39]:
dict_df=pd.DataFrame(dict_olymp)
dict_df

HostCity  Year  No of countries
0      GBR  2012              205
1      CHN  2008              204
2      RUS  2004              201
3       US  2000              200
4      KOR  2006              197

In [40]:
dict_df.HostCity

0    GBR
1    CHN
2    RUS
3     US
4    KOR
Name: HostCity, dtype: object

In [41]:
dict_df['HostCity']

0    GBR
1    CHN
2    RUS
3     US
4    KOR
Name: HostCity, dtype: object

In [42]:
dict_df[['HostCity']]

HostCity
0      GBR
1      CHN
2      RUS
3       US
4      KOR

In [43]:
dict_df.describe

<bound method NDFrame.describe of   HostCity  Year  No of countries
0      GBR  2012              205
1      CHN  2008              204
2      RUS  2004              201
3       US  2000              200
4      KOR  2006              197>

In [45]:
olymp_series=pd.Series([205, 204, 201, 200, 197],index=[2012, 2008, 2004, 2000, 2006])
olymp_seriesc=pd.Series(['GBR', 'CHN', 'RUS', 'US', 'KOR'],index=[2012, 2008, 2004, 2000, 2006])

series_df=pd.DataFrame({'No of countries':olymp_series,'Host Cities':olymp_seriesc})
series_df

No of countries Host Cities
2012              205         GBR
2008              204         CHN
2004              201         RUS
2000              200          US
2006              197         KOR

In [47]:
series_df.describe

<bound method NDFrame.describe of       No of countries Host Cities
2012              205         GBR
2008              204         CHN
2004              201         RUS
2000              200          US
2006              197         KOR>

In [50]:
np_arr=np.array([2012, 2008, 2004, 2000, 2006])
np_year={'year':np_arr}
df_array=pd.DataFrame(np_year)   
df_array

year
0  2012
1  2008
2  2004
3  2000
4  2006

In [51]:
df_from_df=pd.DataFrame(series_df)
df_from_df

No of countries Host Cities
2012              205         GBR
2008              204         CHN
2004              201         RUS
2000              200          US
2006              197         KOR

In [52]:
df_from_df.columns

Index(['No of countries', 'Host Cities'], dtype='object')

In [53]:
df_from_df.index

Int64Index([2012, 2008, 2004, 2000, 2006], dtype='int64')

In [54]:
df_from_df.loc[2012]

No of countries    205
Host Cities        GBR
Name: 2012, dtype: object

In [55]:
df_from_df.iat[3,1]

'US'

In [59]:
df_from_df.iloc[3,1]

'US'

In [57]:
df_from_df.iloc[1:3]

No of countries Host Cities
2008              204         CHN
2004              201         RUS

In [58]:
df_from_df.iloc[3]

No of countries    200
Host Cities         US
Name: 2000, dtype: object

In [62]:
see=df_from_df['No of countries']>200
see

2012     True
2008     True
2004     True
2000    False
2006    False
Name: No of countries, dtype: bool

In [64]:
seed=df_from_df[df_from_df['No of countries']>200]
seed

No of countries Host Cities
2012              205         GBR
2008              204         CHN
2004              201         RUS

## missing values

are treated with dropna() or fillna(), Inplace=True

## Data Operations

In [65]:
df_movie_rating=pd.DataFrame({'movie1':[5,4,3,3,2,1],'movie2':[4,5,2,3,4,2]},
                             index=['tom','jeff','ada','obi','ugu','uja'])
df_movie_rating

movie1  movie2
tom        5       4
jeff       4       5
ada        3       2
obi        3       3
ugu        2       4
uja        1       2

In [94]:
df_movie_rating.shape

(6, 2)

In [95]:
df_movie_rating.size

12

In [96]:
df_movie_rating.ndim

2

In [67]:
def movie_grade(rating):
    if rating==5:
        return 'A'
    elif rating==4:
        return 'B'
    elif rating==3:
        return 'C'
    else:
        return 'F'

In [68]:
movie_grade(4)

'B'

In [71]:
#movie_grade(df_movie_rating) wont work as its a table, instead


df_movie_rating.applymap(movie_grade)

movie1 movie2
tom       A      B
jeff      B      A
ada       C      F
obi       C      C
ugu       F      B
uja       F      F

In [72]:
df_scores=pd.DataFrame({'test1':[95,84,73,88,82,61],'test2':[34,67,89,87,49,88]},
                        index=['tom','jeff','ada','obi','ugu','uja'])
df_scores

test1  test2
tom      95     34
jeff     84     67
ada      73     89
obi      88     87
ugu      82     49
uja      61     88

In [73]:
df_scores.max()

test1    95
test2    89
dtype: int64

In [74]:
df_scores.min()

test1    61
test2    34
dtype: int64

In [75]:
df_scores.std()

test1    11.979149
test2    23.298069
dtype: float64

In [76]:
df_scores.mean()

test1    80.5
test2    69.0
dtype: float64

In [77]:
df_names=pd.DataFrame({'first':['uja','jeff','ada','obi','ugu','uja'],
                       'last':['obu','Emekwo','john','udo','Meg','Uka']},
                        )
df_names

first    last
0   uja     obu
1  jeff  Emekwo
2   ada    john
3   obi     udo
4   ugu     Meg
5   uja     Uka

In [80]:
grouped=df_names.groupby('first')
grouped

In [81]:
grp=grouped.get_group('uja')
grp

first last
0   uja  obu
5   uja  Uka

In [84]:
df_names.sort_values('first')

first    last
2   ada    john
1  jeff  Emekwo
3   obi     udo
4   ugu     Meg
0   uja     obu
5   uja     Uka

## Data  Standardization

In [85]:
def standardize_test(test):
    return (test - test.mean()) / test.std()

In [87]:
standardize_test(df_scores['test1'])

tom     1.210437
jeff    0.292174
ada    -0.626088
obi     0.626088
ugu     0.125218
uja    -1.627829
Name: test1, dtype: float64

In [88]:
standardize_test(df_scores)

test1     test2
tom   1.210437 -1.502270
jeff  0.292174 -0.085844
ada  -0.626088  0.858440
obi   0.626088  0.772596
ugu   0.125218 -0.858440
uja  -1.627829  0.815518

In [92]:
def standardize_test_table(df):
    return df_scores.apply(standardize_test)

In [93]:
standardize_test_table(df_scores)

test1     test2
tom   1.210437 -1.502270
jeff  0.292174 -0.085844
ada  -0.626088  0.858440
obi   0.626088  0.772596
ugu   0.125218 -0.858440
uja  -1.627829  0.815518

## other operations include Merge, join, concat, drop_duplicate()

# Pandas SQL OPeration

In [ ]:
import sqlite3

In [ ]:
#create sql table
create_table="""
CREATE TABLE student_score
(Id INTEGER, Name VARCHAR(20), Math REAL,
Science REAL
);"""

In [ ]:
#execute the SQL statement
executeSQL=sqlite3.connect(':memory:')
executeSQL.execute(create_table)
executeSQl.commit()

In [ ]:
#prepare a SQL query
SQL_query = execute('select * from student-score')

In [ ]:
#fetch result from the sqlite database
resultset = SQL_query.fetchall()
resultset

In [ ]:
#prepare records to be inserted into sql tavble through sql statement
insertSQL = [(10,'Jack',85,89),(23,'Jay',85,86),(13,'Jane'95,39)]

In [ ]:
#insert records into Sql table through SQl statement
insert_statement = 'insert into student_score vales(?,?,?,?)'
executeSQL.executemany(insert_statement,insertSQL)
executeSQL.commit()

In [ ]:
#prepare sql query
resltset=SQL_query.fetchall()

In [ ]:
#view the resultset
resltset

In [ ]:
# put the record together in a dataframe
df_student_rec=pd.DataFrame(resltset,columns=zip(*SQL_query.description)[0])

In [ ]:
#view the df in pandas
df_student_rec

# Data Manipulation

In [150]:
df_car=pd.read_csv('../course3/CARS.csv')
df_car.head()

name_of_car           Model   Type Origin DriveTrain     MSRP  Invoice  \
0       Acura             MDX    SUV   Asia        All  $36,945  $33,337   
1       Acura  RSX Type S 2dr  Sedan   Asia      Front  $23,820  $21,761   
2       Acura         TSX 4dr  Sedan   Asia      Front  $26,990  $24,647   
3       Acura          TL 4dr  Sedan   Asia      Front  $33,195  $30,299   
4       Acura      3.5 RL 4dr  Sedan   Asia      Front  $43,755  $39,014   

   EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  Weight  \
0         3.5        6.0         265        17           23    4451   
1         2.0        4.0         200        24           31    2778   
2         2.4        4.0         200        22           29    3230   
3         3.2        6.0         270        20           28    3575   
4         3.5        6.0         225        18           24    3880   

   Wheelbase  Length  
0        106     189  
1        101     172  
2        105     183  
3        108     186  
4        115     197

In [151]:
df_car.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 428 entries, 0 to 427
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   name_of_car  428 non-null    object 
 1   Model        428 non-null    object 
 2   Type         428 non-null    object 
 3   Origin       428 non-null    object 
 4   DriveTrain   428 non-null    object 
 5   MSRP         428 non-null    object 
 6   Invoice      428 non-null    object 
 7   EngineSize   428 non-null    float64
 8   Cylinders    426 non-null    float64
 9   Horsepower   428 non-null    int64  
 10  MPG_City     428 non-null    int64  
 11  MPG_Highway  428 non-null    int64  
 12  Weight       428 non-null    int64  
 13  Wheelbase    428 non-null    int64  
 14  Length       428 non-null    int64  
dtypes: float64(2), int64(6), object(7)
memory usage: 50.3+ KB


In [152]:
df_car.tail(15)

name_of_car                    Model   Type  Origin DriveTrain     MSRP  \
413  Volkswagen                 Jetta GL  Wagon  Europe      Front  $19,005   
414  Volkswagen          Passat GLS 1.8T  Wagon  Europe      Front  $24,955   
415  Volkswagen                Passat W8  Wagon  Europe      Front  $40,235   
416       Volvo                  XC90 T6    SUV  Europe        All  $41,250   
417       Volvo                  S40 4dr  Sedan  Europe      Front  $25,135   
418       Volvo              S60 2.5 4dr  Sedan  Europe        All  $31,745   
419       Volvo               S60 T5 4dr  Sedan  Europe      Front  $34,845   
420       Volvo                S60 R 4dr  Sedan  Europe        All  $37,560   
421       Volvo              S80 2.9 4dr  Sedan  Europe      Front  $37,730   
422       Volvo             S80 2.5T 4dr  Sedan  Europe        All  $37,885   
423       Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  $40,565   
424       Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  $42,565   
425       Volvo               S80 T6 4dr  Sedan  Europe      Front  $45,210   
426       Volvo                      V40  Wagon  Europe      Front  $26,135   
427       Volvo                     XC70  Wagon  Europe        All  $35,145   

     Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
413  $17,427         2.0        4.0         115        24           30   
414  $22,801         1.8        4.0         170        22           31   
415  $36,956         4.0        8.0         270        18           25   
416  $38,851         2.9        6.0         268        15           20   
417  $23,701         1.9        4.0         170        22           29   
418  $29,916         2.5        5.0         208        20           27   
419  $32,902         2.3        5.0         247        20           28   
420  $35,382         2.5        5.0         300        18           25   
421  $35,542         2.9        6.0         208        20           28   
422  $35,688         2.5        5.0         194        20           27   
423  $38,203         2.4        5.0         197        21           28   
424  $40,083         2.3        5.0         242        20           26   
425  $42,573         2.9        6.0         268        19           26   
426  $24,641         1.9        4.0         170        22           29   
427  $33,112         2.5        5.0         208        20           27   

     Weight  Wheelbase  Length  
413    3034         99     174  
414    3338        106     184  
415    4067        106     184  
416    4638        113     189  
417    2767        101     178  
418    3903        107     180  
419    3766        107     180  
420    3571        107     181  
421    3576        110     190  
422    3691        110     190  
423    3450        105     186  
424    3450        105     186  
425    3653        110     190  
426    2822        101     180  
427    3823        109     186

#to delete unwanted rows using index 


#df_car.drop(range(428,443),inplace=True)
#df_car.drop(df_car.index[428:443])

In [153]:
df_car.tail()

name_of_car                    Model   Type  Origin DriveTrain     MSRP  \
423       Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  $40,565   
424       Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  $42,565   
425       Volvo               S80 T6 4dr  Sedan  Europe      Front  $45,210   
426       Volvo                      V40  Wagon  Europe      Front  $26,135   
427       Volvo                     XC70  Wagon  Europe        All  $35,145   

     Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
423  $38,203         2.4        5.0         197        21           28   
424  $40,083         2.3        5.0         242        20           26   
425  $42,573         2.9        6.0         268        19           26   
426  $24,641         1.9        4.0         170        22           29   
427  $33,112         2.5        5.0         208        20           27   

     Weight  Wheelbase  Length  
423    3450        105     186  
424    3450        105     186  
425    3653        110     190  
426    2822        101     180  
427    3823        109     186

In [154]:
df_car.describe

<bound method NDFrame.describe of     name_of_car                    Model   Type  Origin DriveTrain     MSRP  \
0         Acura                      MDX    SUV    Asia        All  $36,945   
1         Acura           RSX Type S 2dr  Sedan    Asia      Front  $23,820   
2         Acura                  TSX 4dr  Sedan    Asia      Front  $26,990   
3         Acura                   TL 4dr  Sedan    Asia      Front  $33,195   
4         Acura               3.5 RL 4dr  Sedan    Asia      Front  $43,755   
..          ...                      ...    ...     ...        ...      ...   
423       Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  $40,565   
424       Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  $42,565   
425       Volvo               S80 T6 4dr  Sedan  Europe      Front  $45,210   
426       Volvo                      V40  Wagon  Europe      Front  $26,135   
427       Volvo                     XC70  Wagon  Europe        All  $35,145   

     Invoice  Eng

In [155]:
df_car.describe(include='all')

name_of_car     Model   Type Origin DriveTrain     MSRP  Invoice  \
count          428       428    428    428        428      428      428   
unique          38       425      6      3          3      410      425   
top         Toyota  C320 4dr  Sedan   Asia      Front  $21,595  $68,306   
freq            28         2    262    158        226        2        2   
mean           NaN       NaN    NaN    NaN        NaN      NaN      NaN   
std            NaN       NaN    NaN    NaN        NaN      NaN      NaN   
min            NaN       NaN    NaN    NaN        NaN      NaN      NaN   
25%            NaN       NaN    NaN    NaN        NaN      NaN      NaN   
50%            NaN       NaN    NaN    NaN        NaN      NaN      NaN   
75%            NaN       NaN    NaN    NaN        NaN      NaN      NaN   
max            NaN       NaN    NaN    NaN        NaN      NaN      NaN   

        EngineSize   Cylinders  Horsepower    MPG_City  MPG_Highway  \
count   428.000000  426.000000  428.000000  428.000000   428.000000   
unique         NaN         NaN         NaN         NaN          NaN   
top            NaN         NaN         NaN         NaN          NaN   
freq           NaN         NaN         NaN         NaN          NaN   
mean      3.196729    5.807512  215.885514   20.060748    26.843458   
std       1.108595    1.558443   71.836032    5.238218     5.741201   
min       1.300000    3.000000   73.000000   10.000000    12.000000   
25%       2.375000    4.000000  165.000000   17.000000    24.000000   
50%       3.000000    6.000000  210.000000   19.000000    26.000000   
75%       3.900000    6.000000  255.000000   21.250000    29.000000   
max       8.300000   12.000000  500.000000   60.000000    66.000000   

             Weight   Wheelbase      Length  
count    428.000000  428.000000  428.000000  
unique          NaN         NaN         NaN  
top             NaN         NaN         NaN  
freq            NaN         NaN         NaN  
mean    3577.953271  108.154206  186.362150  
std      758.983215    8.311813   14.357991  
min     1850.000000   89.000000  143.000000  
25%     3104.000000  103.000000  178.000000  
50%     3474.500000  107.000000  187.000000  
75%     3977.750000  112.000000  194.000000  
max     7190.000000  144.000000  238.000000

In [156]:
df_car.isnull().sum()

name_of_car    0
Model          0
Type           0
Origin         0
DriveTrain     0
MSRP           0
Invoice        0
EngineSize     0
Cylinders      2
Horsepower     0
MPG_City       0
MPG_Highway    0
Weight         0
Wheelbase      0
Length         0
dtype: int64

In [158]:
df_car['Cylinders'].mean()

5.807511737089202

In [159]:
df_car['Cylinders'].fillna(df_car['Cylinders'].mean(),inplace=True)

In [160]:
df_car.columns

Index(['name_of_car', 'Model', 'Type', 'Origin', 'DriveTrain', 'MSRP',
       'Invoice', 'EngineSize', 'Cylinders', 'Horsepower', 'MPG_City',
       'MPG_Highway', 'Weight', 'Wheelbase', 'Length'],
      dtype='object')

In [161]:
# apply a function on a column of the dataframe
# squaring the cyclinders

# df.apply(function,axis=0(for applying the func over columns, axis=1 (for rows)))


# .apply for applying func to either col or row(series), .applymap for applying func to df

df_car['new_cylinders']=df_car['Cylinders'].apply(lambda x:x**2)

df_car.head()

name_of_car           Model   Type Origin DriveTrain     MSRP  Invoice  \
0       Acura             MDX    SUV   Asia        All  $36,945  $33,337   
1       Acura  RSX Type S 2dr  Sedan   Asia      Front  $23,820  $21,761   
2       Acura         TSX 4dr  Sedan   Asia      Front  $26,990  $24,647   
3       Acura          TL 4dr  Sedan   Asia      Front  $33,195  $30,299   
4       Acura      3.5 RL 4dr  Sedan   Asia      Front  $43,755  $39,014   

   EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  Weight  \
0         3.5        6.0         265        17           23    4451   
1         2.0        4.0         200        24           31    2778   
2         2.4        4.0         200        22           29    3230   
3         3.2        6.0         270        20           28    3575   
4         3.5        6.0         225        18           24    3880   

   Wheelbase  Length  new_cylinders  
0        106     189           36.0  
1        101     172           16.0  
2        105     183           16.0  
3        108     186           36.0  
4        115     197           36.0

In [162]:
df_car.index

RangeIndex(start=0, stop=428, step=1)

In [163]:
df_car.set_index('Model',drop=True,inplace=True)
# to reset index
#df_car.reset_index(inplace=True)

In [164]:
df_car.head(2)

name_of_car   Type Origin DriveTrain     MSRP  Invoice  \
Model                                                                   
MDX                  Acura    SUV   Asia        All  $36,945  $33,337   
RSX Type S 2dr       Acura  Sedan   Asia      Front  $23,820  $21,761   

                EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
Model                                                                      
MDX                    3.5        6.0         265        17           23   
RSX Type S 2dr         2.0        4.0         200        24           31   

                Weight  Wheelbase  Length  new_cylinders  
Model                                                     
MDX               4451        106     189           36.0  
RSX Type S 2dr    2778        101     172           16.0

In [165]:
df_car.dtypes

name_of_car       object
Type              object
Origin            object
DriveTrain        object
MSRP              object
Invoice           object
EngineSize       float64
Cylinders        float64
Horsepower         int64
MPG_City           int64
MPG_Highway        int64
Weight             int64
Wheelbase          int64
Length             int64
new_cylinders    float64
dtype: object

In [166]:
df_car[['Type','Origin']]

Type  Origin
Model                                 
MDX                        SUV    Asia
RSX Type S 2dr           Sedan    Asia
TSX 4dr                  Sedan    Asia
TL 4dr                   Sedan    Asia
3.5 RL 4dr               Sedan    Asia
...                        ...     ...
C70 LPT convertible 2dr  Sedan  Europe
C70 HPT convertible 2dr  Sedan  Europe
S80 T6 4dr               Sedan  Europe
V40                      Wagon  Europe
XC70                     Wagon  Europe

[428 rows x 2 columns]

In [168]:
df_car.iloc[247,]

name_of_car        Mazda
Type              Sports
Origin              Asia
DriveTrain          Rear
MSRP             $25,700
Invoice          $23,794
EngineSize           1.3
Cylinders        5.80751
Horsepower           197
MPG_City              18
MPG_Highway           25
Weight              3053
Wheelbase            106
Length               174
new_cylinders    33.7272
Name: RX-8 4dr automatic, dtype: object

In [169]:
df_car.iloc[247,6]

1.3

In [170]:
df_car.iloc[247,4:8]

MSRP          $25,700
Invoice       $23,794
EngineSize        1.3
Cylinders     5.80751
Name: RX-8 4dr automatic, dtype: object

In [175]:
df_car.iloc[247:250,:5]

name_of_car    Type Origin DriveTrain     MSRP
Model                                                              
RX-8 4dr automatic         Mazda  Sports   Asia       Rear  $25,700
RX-8 4dr manual            Mazda  Sports   Asia       Rear  $27,200
B2300 SX Regular Cab       Mazda   Truck   Asia       Rear  $14,840

In [177]:
df_car.loc['MDX':'3.5 RL 4dr','Type':'Origin']

Type Origin
Model                       
MDX               SUV   Asia
RSX Type S 2dr  Sedan   Asia
TSX 4dr         Sedan   Asia
TL 4dr          Sedan   Asia
3.5 RL 4dr      Sedan   Asia

In [191]:
df_car['Origin'].nunique()

3

In [189]:
df_car['Origin'].value_counts()

Asia      158
USA       147
Europe    123
Name: Origin, dtype: int64

In [192]:
df_car['Origin'].unique()

array(['Asia', 'Europe', 'USA'], dtype=object)

In [217]:
# boolean indexing
# extract all the cars where origin =asia
df_car[df_car['Origin']=='USA'].head(3)


name_of_car   Type Origin DriveTrain     MSRP  Invoice  \
Model                                                                       
Rainier                  Buick    SUV    USA        All  $37,895  $34,357   
Rendezvous CX            Buick    SUV    USA      Front  $26,545  $24,085   
Century Custom 4dr       Buick  Sedan    USA      Front  $22,180  $20,351   

                    EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
Model                                                                          
Rainier                    4.2        6.0         275        15           21   
Rendezvous CX              3.4        6.0         185        19           26   
Century Custom 4dr         3.1        6.0         175        20           30   

                    Weight  Wheelbase  Length  new_cylinders  
Model                                                         
Rainier               4600        113     193           36.0  
Rendezvous CX         4024        112     187           36.0  
Century Custom 4dr    3353        109     195           36.0

In [220]:
df_car[(df_car['Type']=='SUV')&(df_car['Origin']=='Asia')].head()

name_of_car Type Origin DriveTrain     MSRP  Invoice  EngineSize  \
Model                                                                           
MDX                Acura  SUV   Asia        All  $36,945  $33,337         3.5   
Pilot LX           Honda  SUV   Asia        All  $27,560  $24,843         3.5   
CR-V LX            Honda  SUV   Asia        All  $19,860  $18,419         2.4   
Element LX         Honda  SUV   Asia        All  $18,690  $17,334         2.4   
Santa Fe GLS     Hyundai  SUV   Asia      Front  $21,589  $20,201         2.7   

              Cylinders  Horsepower  MPG_City  MPG_Highway  Weight  Wheelbase  \
Model                                                                           
MDX                 6.0         265        17           23    4451        106   
Pilot LX            6.0         240        17           22    4387        106   
CR-V LX             4.0         160        21           25    3258        103   
Element LX          4.0         160        21           24    3468        101   
Santa Fe GLS        6.0         173        20           26    3549        103   

              Length  new_cylinders  
Model                                
MDX              189           36.0  
Pilot LX         188           36.0  
CR-V LX          179           16.0  
Element LX       167           16.0  
Santa Fe GLS     177           36.0

In [221]:
df_car[(df_car['Type']=='SUV')|(df_car['Origin']=='Asia')].head()

name_of_car   Type Origin DriveTrain     MSRP  Invoice  \
Model                                                                   
MDX                  Acura    SUV   Asia        All  $36,945  $33,337   
RSX Type S 2dr       Acura  Sedan   Asia      Front  $23,820  $21,761   
TSX 4dr              Acura  Sedan   Asia      Front  $26,990  $24,647   
TL 4dr               Acura  Sedan   Asia      Front  $33,195  $30,299   
3.5 RL 4dr           Acura  Sedan   Asia      Front  $43,755  $39,014   

                EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
Model                                                                      
MDX                    3.5        6.0         265        17           23   
RSX Type S 2dr         2.0        4.0         200        24           31   
TSX 4dr                2.4        4.0         200        22           29   
TL 4dr                 3.2        6.0         270        20           28   
3.5 RL 4dr             3.5        6.0         225        18           24   

                Weight  Wheelbase  Length  new_cylinders  
Model                                                     
MDX               4451        106     189           36.0  
RSX Type S 2dr    2778        101     172           16.0  
TSX 4dr           3230        105     183           16.0  
TL 4dr            3575        108     186           36.0  
3.5 RL 4dr        3880        115     197           36.0

In [223]:
#sort by horsepower in ascending order
df_car.sort_values(by='Horsepower').head(3)

name_of_car    Type  Origin DriveTrain  \
Model                                                                           
Insight 2dr (gas/electric)                   Honda  Hybrid    Asia      Front   
Civic Hybrid 4dr manual (gas/electric)       Honda  Hybrid    Asia      Front   
Jetta GLS TDI 4dr                       Volkswagen   Sedan  Europe      Front   

                                           MSRP  Invoice  EngineSize  \
Model                                                                  
Insight 2dr (gas/electric)              $19,110  $17,911         2.0   
Civic Hybrid 4dr manual (gas/electric)  $20,140  $18,451         1.4   
Jetta GLS TDI 4dr                       $21,055  $19,638         1.9   

                                        Cylinders  Horsepower  MPG_City  \
Model                                                                     
Insight 2dr (gas/electric)                    3.0          73        60   
Civic Hybrid 4dr manual (gas/electric)        4.0          93        46   
Jetta GLS TDI 4dr                             4.0         100        38   

                                        MPG_Highway  Weight  Wheelbase  \
Model                                                                    
Insight 2dr (gas/electric)                       66    1850         95   
Civic Hybrid 4dr manual (gas/electric)           51    2732        103   
Jetta GLS TDI 4dr                                46    3003         99   

                                        Length  new_cylinders  
Model                                                          
Insight 2dr (gas/electric)                 155            9.0  
Civic Hybrid 4dr manual (gas/electric)     175           16.0  
Jetta GLS TDI 4dr                          172           16.0

In [227]:
#in descending order
df_car.sort_values(by='MSRP',ascending=False).head(3)

name_of_car    Type  Origin DriveTrain     MSRP  \
Model                                                                       
CL500 2dr               Mercedes-Benz   Sedan  Europe       Rear  $94,820   
SL500 convertible 2dr   Mercedes-Benz  Sports  Europe       Rear  $90,520   
NSX coupe 2dr manual S          Acura  Sports    Asia       Rear  $89,765   

                        Invoice  EngineSize  Cylinders  Horsepower  MPG_City  \
Model                                                                          
CL500 2dr               $88,324         5.0        8.0         302        16   
SL500 convertible 2dr   $84,325         5.0        8.0         302        16   
NSX coupe 2dr manual S  $79,978         3.2        6.0         290        17   

                        MPG_Highway  Weight  Wheelbase  Length  new_cylinders  
Model                                                                          
CL500 2dr                        24    4085        114     196           64.0  
SL500 convertible 2dr            23    4065        101     179           64.0  
NSX coupe 2dr manual S           24    3153        100     174           36.0

In [229]:
df_car.sort_values(by=['EngineSize','Cylinders'],ascending=[False,True]).head(3)

name_of_car    Type Origin DriveTrain     MSRP  \
Model                                                                         
Viper SRT-10 convertible 2dr       Dodge  Sports    USA       Rear  $81,795   
Excursion 6.8 XLT                   Ford     SUV    USA        All  $41,475   
Escalade EXT                    Cadillac   Truck    USA        All  $52,975   

                              Invoice  EngineSize  Cylinders  Horsepower  \
Model                                                                      
Viper SRT-10 convertible 2dr  $74,451         8.3       10.0         500   
Excursion 6.8 XLT             $36,494         6.8       10.0         310   
Escalade EXT                  $48,541         6.0        8.0         345   

                              MPG_City  MPG_Highway  Weight  Wheelbase  \
Model                                                                    
Viper SRT-10 convertible 2dr        12           20    3410         99   
Excursion 6.8 XLT                   10           13    7190        137   
Escalade EXT                        13           17    5879        130   

                              Length  new_cylinders  
Model                                                
Viper SRT-10 convertible 2dr     176          100.0  
Excursion 6.8 XLT                227          100.0  
Escalade EXT                     221           64.0

In [261]:
df_car['Type'].value_counts()

Sedan     262
SUV        60
Sports     49
Wagon      30
Truck      24
Hybrid      3
Name: Type, dtype: int64

In [237]:
pd.crosstab(index=df_car['Type'],columns='no_of_types')

col_0   no_of_types
Type               
Hybrid            3
SUV              60
Sedan           262
Sports           49
Truck            24
Wagon            30

In [257]:
pd.crosstab(index=df_car['Type'],columns=df_car['Origin'],margins=True)

Origin  Asia  Europe  USA  All
Type                          
Hybrid     3       0    0    3
SUV       25      10   25   60
Sedan     94      78   90  262
Sports    17      23    9   49
Truck      8       0   16   24
Wagon     11      12    7   30
All      158     123  147  428

In [266]:
# pivot table 


#df. pivot_table(index=rows,columns=,aggfunc=np.mean/sum/count)

#average Horsepower by the Type of the car
df_car.pivot_table(index=('Type'),values=('Horsepower'),aggfunc='mean')

Horsepower
Type              
Hybrid   92.000000
SUV     235.816667
Sedan   201.656489
Sports  284.163265
Truck   224.833333
Wagon   194.000000

In [268]:
#average Horsepower by the Type of the car and the origin
df_car.pivot_table(index='Type',columns='Origin',values='Horsepower',aggfunc='mean')

Origin        Asia      Europe         USA
Type                                      
Hybrid   92.000000         NaN         NaN
SUV     214.160000  263.100000  246.560000
Sedan   181.978723  236.525641  191.988889
Sports  225.352941  316.739130  312.000000
Truck   190.250000         NaN  242.125000
Wagon   185.636364  218.166667  165.714286

In [280]:
# average Horsepower by the Type

# car.groupby(variable to group by)
Typgrp=df_car.groupby(df_car['Type'])
Typgrp

In [278]:
Typgrp.get_group('Sedan')

name_of_car   Type  Origin DriveTrain     MSRP  \
Model                                                                    
RSX Type S 2dr                Acura  Sedan    Asia      Front  $23,820   
TSX 4dr                       Acura  Sedan    Asia      Front  $26,990   
TL 4dr                        Acura  Sedan    Asia      Front  $33,195   
3.5 RL 4dr                    Acura  Sedan    Asia      Front  $43,755   
3.5 RL w/Navigation 4dr       Acura  Sedan    Asia      Front  $46,100   
...                             ...    ...     ...        ...      ...   
S80 2.9 4dr                   Volvo  Sedan  Europe      Front  $37,730   
S80 2.5T 4dr                  Volvo  Sedan  Europe        All  $37,885   
C70 LPT convertible 2dr       Volvo  Sedan  Europe      Front  $40,565   
C70 HPT convertible 2dr       Volvo  Sedan  Europe      Front  $42,565   
S80 T6 4dr                    Volvo  Sedan  Europe      Front  $45,210   

                         Invoice  EngineSize  Cylinders  Horsepower  MPG_City  \
Model                                                                           
RSX Type S 2dr           $21,761         2.0        4.0         200        24   
TSX 4dr                  $24,647         2.4        4.0         200        22   
TL 4dr                   $30,299         3.2        6.0         270        20   
3.5 RL 4dr               $39,014         3.5        6.0         225        18   
3.5 RL w/Navigation 4dr  $41,100         3.5        6.0         225        18   
...                          ...         ...        ...         ...       ...   
S80 2.9 4dr              $35,542         2.9        6.0         208        20   
S80 2.5T 4dr             $35,688         2.5        5.0         194        20   
C70 LPT convertible 2dr  $38,203         2.4        5.0         197        21   
C70 HPT convertible 2dr  $40,083         2.3        5.0         242        20   
S80 T6 4dr               $42,573         2.9        6.0         268        19   

                         MPG_Highway  Weight  Wheelbase  Length  new_cylinders  
Model                                                                           
RSX Type S 2dr                    31    2778        101     172           16.0  
TSX 4dr                           29    3230        105     183           16.0  
TL 4dr                            28    3575        108     186           36.0  
3.5 RL 4dr                        24    3880        115     197           36.0  
3.5 RL w/Navigation 4dr           24    3893        115     197           36.0  
...                              ...     ...        ...     ...            ...  
S80 2.9 4dr                       28    3576        110     190           36.0  
S80 2.5T 4dr                      27    3691        110     190           25.0  
C70 LPT convertible 2dr           28    3450        105     186           25.0  
C70 HPT convertible 2dr           26    3450        105     186           25.0  
S80 T6 4dr                        26    3653        110     190           36.0  

[262 rows x 15 columns]

In [281]:
Typgrp.groups

{'Hybrid': Index(['Civic Hybrid 4dr manual (gas/electric)', 'Insight 2dr (gas/electric)',
        'Prius 4dr (gas/electric)'],
       dtype='object', name='Model'),
 'SUV': Index(['MDX', 'X3 3.0i', 'X5 4.4i', 'Rainier', 'Rendezvous CX', 'Escalade',
        'SRX V8', 'Suburban 1500 LT', 'Tahoe LT', 'TrailBlazer LT', 'Tracker',
        'Durango SLT', 'Excursion 6.8 XLT', 'Expedition 4.6 XLT',
        'Explorer XLT V6', 'Escape XLS', 'Envoy XUV SLE', 'Yukon 1500 SLE',
        'Yukon XL 2500 SLT', 'Pilot LX', 'CR-V LX', 'Element LX', 'H2',
        'Santa Fe GLS', 'Ascender S', 'Rodeo S', 'Grand Cherokee Laredo',
        'Liberty Sport', 'Wrangler Sahara convertible 2dr', 'Sorento LX',
        'Range Rover HSE', 'Discovery SE', 'Freelander SE', 'GX 470', 'LX 470',
        'RX 330', 'Navigator Luxury', 'Aviator Ultimate', 'Tribute DX 2.0',
        'G500', 'ML500', 'Mountaineer', 'Endeavor XLS', 'Montero XLS',
        'Outlander LS', 'Pathfinder Armada SE', 'Pathfinder SE', 'Xterra XE V6',
  

In [283]:
for typ,name in Typgrp:
    print(typ)
    print(name)

Hybrid
                                       name_of_car    Type Origin DriveTrain  \
Model                                                                          
Civic Hybrid 4dr manual (gas/electric)       Honda  Hybrid   Asia      Front   
Insight 2dr (gas/electric)                   Honda  Hybrid   Asia      Front   
Prius 4dr (gas/electric)                    Toyota  Hybrid   Asia      Front   

                                           MSRP  Invoice  EngineSize  \
Model                                                                  
Civic Hybrid 4dr manual (gas/electric)  $20,140  $18,451         1.4   
Insight 2dr (gas/electric)              $19,110  $17,911         2.0   
Prius 4dr (gas/electric)                $20,510  $18,926         1.5   

                                        Cylinders  Horsepower  MPG_City  \
Model                                                                     
Civic Hybrid 4dr manual (gas/electric)        4.0          93        46   
Insigh

In [285]:
df_car.groupby(df_car['Type'])['Horsepower']

In [286]:
df_car.groupby(df_car['Type'])['Horsepower'].mean()

Type
Hybrid     92.000000
SUV       235.816667
Sedan     201.656489
Sports    284.163265
Truck     224.833333
Wagon     194.000000
Name: Horsepower, dtype: float64

In [287]:
#max enginesize by origin
df_car.groupby(df_car['Origin'])['EngineSize'].max()

Origin
Asia      5.6
Europe    6.0
USA       8.3
Name: EngineSize, dtype: float64

In [294]:
df_car.groupby([df_car['Origin'],df_car['Type']])['Horsepower'].agg(['mean','sum','max','min','count'])

mean    sum  max  min  count
Origin Type                                      
Asia   Hybrid   92.000000    276  110   73      3
       SUV     214.160000   5354  325  130     25
       Sedan   181.978723  17106  340  103     94
       Sports  225.352941   3831  300  138     17
       Truck   190.250000   1522  305  142      8
       Wagon   185.636364   2042  315  104     11
Europe SUV     263.100000   2631  340  174     10
       Sedan   236.525641  18449  493  100     78
       Sports  316.739130   7285  493  180     23
       Wagon   218.166667   2618  340  115     12
USA    SUV     246.560000   6164  325  143     25
       Sedan   191.988889  17279  302  103     90
       Sports  312.000000   2808  500  193      9
       Truck   242.125000   3874  345  143     16
       Wagon   165.714286   1160  250  130      7

In [296]:
df_car.agg({'Weight': max,'Horsepower':min})

Weight        7190
Horsepower      73
dtype: int64

In [297]:
dict={'name':['A','A','A','B','B','C'],'marks':[10,10,20,20,30,40]}
df=pd.DataFrame(dict,index=['p','q','r','s','t','u'])
df

name  marks
p    A     10
q    A     10
r    A     20
s    B     20
t    B     30
u    C     40

In [299]:
# removing the duplicates

# all dups across columns will be removed by index

df.drop_duplicates()

name  marks
p    A     10
r    A     20
s    B     20
t    B     30
u    C     40

In [300]:
df.drop_duplicates(subset='marks')#drops duplictes by marks

name  marks
p    A     10
r    A     20
t    B     30
u    C     40

In [301]:
# drops dups by marks and keeps the last occurence
df.drop_duplicates(subset='marks',keep='last')

name  marks
q    A     10
s    B     20
t    B     30
u    C     40

In [318]:
dict={'name':['A','A','A','B','B','C'],'marks':[10,None,20,None,30,40]}
df=pd.DataFrame(dict,index=['p','q','r','s','t','u'])
df

name  marks
p    A   10.0
q    A    NaN
r    A   20.0
s    B    NaN
t    B   30.0
u    C   40.0

In [303]:
df.isnull()

name  marks
p  False  False
q  False   True
r  False  False
s  False   True
t  False  False
u  False  False

In [304]:
df.isnull().sum()

name     0
marks    2
dtype: int64

In [305]:
df


name  marks
p    A   10.0
q    A    NaN
r    A   20.0
s    B    NaN
t    B   30.0
u    C   40.0

In [309]:
df.nunique()

name     3
marks    4
dtype: int64

In [311]:
df.fillna(method='ffill')#forwardfill

name  marks
p    A   10.0
q    A   10.0
r    A   20.0
s    B   20.0
t    B   30.0
u    C   40.0

In [312]:
df.fillna(method='bfill')#backfill

name  marks
p    A   10.0
q    A   20.0
r    A   20.0
s    B   30.0
t    B   30.0
u    C   40.0

In [321]:
mean_d=df['marks'].mean()
mean_d

25.0

In [313]:
df

name  marks
p    A   10.0
q    A    NaN
r    A   20.0
s    B    NaN
t    B   30.0
u    C   40.0

In [325]:
#this method replaces the corresponding row where there is Nan with df[mean]
df['new']=np.where(df['marks'].isnull(),df['marks'].mean(),df['marks'])
df

name  marks   new
p    A   10.0  10.0
q    A    NaN  25.0
r    A   20.0  20.0
s    B    NaN  25.0
t    B   30.0  30.0
u    C   40.0  40.0

### Scenerio 2

In [17]:
#Import the required libraries
import pandas as pd
import numpy as np

In [18]:
#Import the Fire Department of New York City (FDNY) file
df_fire=pd.read_csv('../dataset/FDNY.csv')

In [19]:
#View the first five records
df_fire.head()

FacilityName       FacilityAddress  \
0                                       FacilityName       FacilityAddress   
1                                 Engine 4/Ladder 15       42 South Street   
2                                Engine 10/Ladder 10    124 Liberty Street   
3                                           Engine 6     49 Beekman Street   
4  Engine 7/Ladder 1/Battalion 1/Manhattan Boroug...  100-104 Duane Street   

     Borough  
0    Borough  
1  Manhattan  
2  Manhattan  
3  Manhattan  
4  Manhattan

In [20]:
#Skip the duplicate header row
df_fire=pd.read_csv('../dataset/FDNY.csv',skiprows=1)

In [21]:
#Verify if the dataset is fixed
df_fire.head()


FacilityName        FacilityAddress  \
0                                 Engine 4/Ladder 15        42 South Street   
1                                Engine 10/Ladder 10     124 Liberty Street   
2                                           Engine 6      49 Beekman Street   
3  Engine 7/Ladder 1/Battalion 1/Manhattan Boroug...   100-104 Duane Street   
4                                           Ladder 8  14 North Moore Street   

     Borough  
0  Manhattan  
1  Manhattan  
2  Manhattan  
3  Manhattan  
4  Manhattan

In [22]:
#View the attributes of the dataset (Hint: view the column names)
df_fire.columns

Index(['FacilityName', 'FacilityAddress', 'Borough'], dtype='object')

In [23]:
df_fire.index

RangeIndex(start=0, stop=218, step=1)

In [24]:
#Count number of records for each attribute
df_fire.count()

FacilityName       218
FacilityAddress    218
Borough            218
dtype: int64

In [25]:
df_fire.dtypes

FacilityName       object
FacilityAddress    object
Borough            object
dtype: object

In [27]:
#Select FDNY information boroughwise
bor=df_fire.groupby('Borough')
bor.size()

Borough
Bronx            34
Brooklyn         66
Manhattan        48
Queens           50
Staten Island    20
dtype: int64

Find the total number of fire department facilities in Manhattan

In [28]:
#Select FDNY information for Manhattan
Manh=bor.get_group('Manhattan')

In [29]:
Manh

FacilityName  \
0                                  Engine 4/Ladder 15   
1                                 Engine 10/Ladder 10   
2                                            Engine 6   
3   Engine 7/Ladder 1/Battalion 1/Manhattan Boroug...   
4                                            Ladder 8   
5                                   Engine 9/Ladder 6   
6                     Engine 15/Ladder 18/Battalion 4   
7                                 Engine 28/Ladder 11   
8                                            Engine 5   
9                                           Engine 55   
10                               Ladder 20/Division 1   
11                     Engine 24/Ladder 5/Battalion 2   
12                                 Engine 33/Ladder 9   
13                               Ladder 3/Battalion 6   
14                                           Squad 18   
15                                Engine 34/Ladder 21   
16                                          Engine 26   
17                     Engine 3/Ladder 12/Battalion 7   
18                                 Engine 1/Ladder 24   
19                                          Engine 14   
20                              Engine 16 / Ladder 7    
21                                          Engine 21   
22                     Engine 54/Ladder 4/Battalion 9   
23                                          Engine 23   
24                                           Rescue 1   
25                                Engine 40/Ladder 35   
26             Ladder 25/District Office 4/Division 3   
27                                          Engine 74   
28                                          Engine 65   
29                  Engine 8 / Ladder 2 / Battalion 8   
30                                Engine 39/Ladder 16   
31                                          Engine 44   
32                   Engine 22/Ladder 13/Battalion 10   
33                                Engine 58/Ladder 26   
34                                Engine 53/Ladder 43   
35                                          Engine 91   
36                                Engine 35/Ladder 14   
37                   Engine 76/Ladder 22/Battalion 11   
38                                          Engine 47   
39                                Engine 59/Ladder 30   
40                                Engine 37/Ladder 40   
41                   Engine 69/Ladder 28/Battalion 16   
42                                Engine 80/Ladder 23   
43                                Engine 84/Ladder 34   
44                                          Engine 67   
45                   Engine 93/Ladder 45/Battalion 13   
46                                Engine 95/Ladder 36   
47                                           Marine 1   

                         FacilityAddress    Borough  
0                        42 South Street  Manhattan  
1                     124 Liberty Street  Manhattan  
2                      49 Beekman Street  Manhattan  
3                   100-104 Duane Street  Manhattan  
4                  14 North Moore Street  Manhattan  
5                        75 Canal Street  Manhattan  
6                         25 Pitt Street  Manhattan  
7                    222 East 2nd Street  Manhattan  
8                   340 East 14th Street  Manhattan  
9                      363 Broome Street  Manhattan  
10                  253 Lafayette Street  Manhattan  
11                     227-29 6th Avenue  Manhattan  
12                 42 Great Jones Street  Manhattan  
13                  108 East 13th Street  Manhattan  
14                  132 West 10th Street  Manhattan  
15                  440 West 38th Street  Manhattan  
16                  220 West 37th Street  Manhattan  
17                  150 West 19th Street  Manhattan  
18               142-46 West 31st Street  Manhattan  
19                   14 East 18th Street  Manhattan  
20                  234 East 29th Street  Manhattan  
21                  238 East 40th Street  Manhattan  
22   